src/toolkits/object_detection/one_shot_object_detection/one_shot_object_detector.cpp

```diff
@@ -63,181 +62,208 @@ flex_dict build_annotation( ParameterSampler &parameter_sampler,
-                                              (background_height-object.m_height)/2);
-
-      flex_dict annotation = build_annotation(parameter_sampler, label,
-                                              object.m_width, object.m_height,
-                                              seed+row_number);
-
-      DASSERT_TRUE(flex_background.get_image_data() != nullptr);
-      DASSERT_TRUE(object.is_decoded());
-      DASSERT_TRUE(flex_background.is_decoded());
-
-      boost::gil::rgba8_image_t::view_t starter_image_view = create_starter_image_view(object);
-
-      boost::gil::rgb8_image_t::view_t background_view = interleaved_view(
-        background_width,
-        background_height,
-        (boost::gil::rgb8_pixel_t*)(flex_background.get_image_data()),
-        background_channels * background_width // row length in bytes
-        );
-
-      images.push_back(
-        create_synthetic_image(starter_image_view, background_view, parameter_sampler, object)
-      );
-      annotations.push_back(annotation);
-
-      if (verbose) {
-        std::ostringstream d;
-        // For pretty printing, floor percent done
-        // resolution to the nearest .25% interval.  Do this by multiplying by
-        // 400, then do integer division by the total size, then float divide
-        // by 4.0
-        int augmented_rows_completed = (input_row_index * backgrounds_size) + row_number;
-        d << augmented_rows_completed * 400 / total_augmented_rows / 4.0 << '%';
-        table.print_progress_row(augmented_rows_completed,
-                                  augmented_rows_completed,
-                                  progress_time(), d.str());
+      for (const auto &row : decompressed_data.range_iterator()) {
+        // go through all the starter images and create augmented images for
+        // all starter images and the respective chunk of background images
+        flex_image object = row[image_column_index].to<flex_image>();
+        std::string label = row[target_column_index].to<flex_string>();
+        std::pair<flex_image, flex_dict> synthetic_row =
+            create_synthetic_image_from_background_and_starter(
+                object, flex_background, label, seed, row_number);
+        flex_image synthetic_image = synthetic_row.first;
+        flex_dict annotation = synthetic_row.second;
+        // write the synthetically generated image and the constructed
+        // annotation to output SFrame.
+        output_writer.write({synthetic_image, annotation}, segment_id);
+        augmented_counter.fetch_add(1, std::memory_order_relaxed);
+        if (verbose) {
+          std::ostringstream d;
+          // For pretty printing, floor percent done
+          // resolution to the nearest .25% interval.  Do this by multiplying by
+          // 400, then do integer division by the total size, then float divide
+          // by 4.0
+          int augmented_rows_completed = augmented_counter;
```

Comment:
You're still reading from the atomic in a separate statement from when you increment it. The primary benefit of `std::atomic` is that you can both increment and read the value at once, atomically

Reply:
Should be fixed now.